agent/containermetadata/metadata_manager.go

```diff
@@ -0,0 +1,186 @@
+	UpdateMetadata(string, *api.Task, *api.Container) error
+	CleanTaskMetadata(*api.Task) error
+}
+
+// metadataManager implements the MetadataManager interface
+type metadataManager struct {
+	client               dockerMetadataClient
+	cluster              string
+	enabled              bool
+	dataDir              string
+	dataDirOnHost        string
+	containerInstanceARN string
+}
+
+// NewMetadataManager creates a metadataManager for a given DockerTaskEngine settings.
+func NewMetadataManager(client dockerMetadataClient, cfg *config.Config) MetadataManager {
+	return &metadataManager{
+		client:        client,
+		cluster:       cfg.Cluster,
+		enabled:       cfg.ContainerMetadataEnabled,
+		dataDir:       cfg.DataDir,
+		dataDirOnHost: cfg.DataDirOnHost,
+	}
+}
+
+// SetContainerInstanceArn sets the metadataManager's ContainerInstanceArn which is not available
+// at its creation as this information is not present immediately at the agent's startup
+func (manager *metadataManager) SetContainerInstanceArn(containerInstanceARN string) {
+	// Do nothing if disabled
+	if manager == nil || !manager.enabled {
+		return
+	}
+	manager.containerInstanceARN = containerInstanceARN
+}
+
+// createmetadata creates the metadata file and adds the metadata directory to
+// the container's mounted host volumes
+// Pointer hostConfig is modified directly so there is risk of concurrency errors.
+func (manager *metadataManager) CreateMetadata(hostConfig *docker.HostConfig, task *api.Task, container *api.Container) error {
+	// Do nothing if disabled
+	if manager == nil || !manager.enabled {
+		return nil
+	}
+
+	// Do not create metadata file for internal containers
+	if container.IsInternal {
+		return nil
+	}
+
+	// Create task and container directories if they do not yet exist
+	metadataDirectoryPath, err := getMetadataFilePath(task, container, manager.dataDir)
+	// Stop metadata creation if path is malformed for any reason
+	if err != nil {
+		return fmt.Errorf("container metadata create: %v", err)
+	}
+
+	err = os.MkdirAll(metadataDirectoryPath, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	// Create metadata file
+	err = createMetadataFile(metadataDirectoryPath)
+	if err != nil {
+		return err
+	}
+
+	// Acquire the metadata then write it in JSON format to the file
+	metadata := manager.parseMetadataAtContainerCreate(task, container)
+	err = metadata.writeToMetadataFile(task, container, manager.dataDir)
+	if err != nil {
+		return err
+	}
+
+	// Add the directory of this container's metadata to the container's mount binds
+	// We do this at the end so that we only mount the directory if there are no errors
+	// This is the only operating system specific point here, so it would be nice if there
+	// were some elegant way to do this for both windows and linux at the same time
+	binds := createBinds(hostConfig.Binds, manager.dataDirOnHost, metadataDirectoryPath, mountPoint, container.Name)
```

Comment:
Yes, that is the plan. However, this will probably require me to rethink how I do the file path parsing.